google/cloud/bigquery/v2/minimal/internal/json_utils.cc

```diff
@@ -106,15 +119,11 @@ nlohmann::json RemoveJsonKeysAndEmptyFields(
               [&parsed](std::string const& key) { return parsed == key; });
           return !discard;
         }
-        if (event == nlohmann::json::parse_event_t::object_end) {
-          return parsed != nullptr && !parsed.empty();
-        }
-        if (event == nlohmann::json::parse_event_t::array_end) {
-          return !parsed.empty();
-        }
```

Comment:
It seems that all you need is to capture a `counters = std::vector<int>{{0}}` value in the `remove_empty_call_back` (nit: should be `_callback`) and then do something like:

```cc
  if (event == nlohmann::json::parse_event_t::object_start || event == nlohmann::json::parse_event_t::array_start) {
    counters.push_back(0);
    return true;
  }
  if (event == nlohmann::json::parse_event_t::key) {
    auto keep = std::all_of(keys.begin(), keys.end(), [&parsed](auto const& k) { return parsed != key; });
    if (keep) ++counters.back();
    return keep;
  }
  if (event == nlohmann::json::parse_event_t::object_end || event == nlohman::json::parse_event_t::array_end) {
    auto count = counters.back();
    counters.pop_back();
    return count != 0;
  }
```

Reply:
I tried such callback function:

```
if (event == nlohmann::json::parse_event_t::array_start) {
          return true; // false - I tried both options
        }
        if (event == nlohmann::json::parse_event_t::array_end) {
          return false;
        }
        return true;
```

works this way. Input: `{"query_parameters":[]}`, output `{"query_parameters":<discarded>}` - which is invalid json. So I don't see any way to remove empty arrays while parsing the string.

Reply:
Thanks for trying that out.